### PR TITLE
Added support for NoneOperation to the control flow graph.

### PR DESCRIPTION
**Pull request by 333fred**
I believe I've covered the interesting cases for expression and statement based NoneOperation instances that can actually cause stack spilling.

### CODE REVIEW

**Comment by 333fred** [comment]
@AlekseyTs @cston @dotnet/roslyn-compiler @dotnet/analyzer-ioperation please review.

**Comment by AlekseyTs** [comment]
@333fred It looks like there are legitimate test failures.

**Comment by 333fred** [comment]
@AlekseyTs, yeah just fixed it.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >captureIdForResult [](start = 96, length = 18)

This is incorrect. The result of the operation is the UnaryOperatorExpression, not its operand. Therefore, we should not be propagating  ```captureIdForResult``` when we are visiting the operand. It sis supposed to be used only for the final result.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >captureIdForResult [](start = 45, length = 18)

Same problem as above

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >int? captureIdForResult [](start = 78, length = 23)

I think this parameter should be removed. Only operations that themselves represent a control flow are allowed to use this value and only for the final result.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >numChildren++ [](start = 16, length = 13)

Alternatively we can simply store the current size of the ```_evalStack``` and then pop according to that. No need to increment inside the loop

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >() => ImmutableArray<IOperation>.Empty [](start = 117, length = 38)

I think we need to make non-lazy version of OperationNone.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >ImmutableArray.CreateBuilder<IOperation>(numChildren) [](start = 34, length = 53)

```ArrayBuilder<IOperation>.GetInstance(numChildren)```

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >childrenBuilder.ToImmutable() [](start = 119, length = 29)

Should be freed, once switched to a pullable builder.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IFixedStatement.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IFixedStatement.cs]: >System.Console.WriteLine($""P is {p}"") [](start = 12, length = 39)

Not required to make changes to the existing tests, but I would recommend to use very simple expressions in order to keep base-lines small. A simple assignment would work, for example.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INoneOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INoneOperation.cs]: >ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1') [](start = 20, length = 88)

I would expect this to be captured too, but we are not handling invocations yet.

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_INoneOperation.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_INoneOperation.vb]: >XXXX [](start = 19, length = 4)

A better name?

**Review by AlekseyTs — DISMISSED** [review]
Done with review pass (iteration 2)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_INoneOperation.vb
Comment (333fred) [reply src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_INoneOperation.vb]: Whoops

---
In reply to: [168012251](https://github.com/dotnet/roslyn/pull/24792#discussion_r168012251) [](ancestors = 168012251)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INoneOperation.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INoneOperation.cs]: I'll add a prototype comment so we track it.

---
In reply to: [168008354](https://github.com/dotnet/roslyn/pull/24792#discussion_r168008354) [](ancestors = 168008354)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IFixedStatement.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IFixedStatement.cs]: Noted.

---
In reply to: [168006425](https://github.com/dotnet/roslyn/pull/24792#discussion_r168006425) [](ancestors = 168006425)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

File: src/Compilers/Core/Portable/Operations/Operation.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/Operation.cs]: >yield return SetParentOperation(child, this); [](start = 24, length = 45)

We should not be doing this for non-lazy flavor. Parent should be assigned to children in constructor. We got rid of the Impl pattern everywhere else

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 4)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @dotnet/analyzer-ioperation ping for reviews.

File: src/Compilers/Core/Portable/Operations/Operation.cs
Comment (ivanbasov) [inline src/Compilers/Core/Portable/Operations/Operation.cs]: nit: extra line

**Review by ivanbasov — APPROVED** [review]
:shipit:

**Comment by 333fred** [comment]
@cston @dotnet/roslyn-compiler I need a second review please.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Are there any tests where `numChildren != operation.Children.Length`?

File: src/Compilers/Core/Portable/Operations/Operation.cs
Comment (cston) [inline src/Compilers/Core/Portable/Operations/Operation.cs]: `readonly`

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Talked offline. This can't happen, I'll add assert to verify that. We aren't just using operation.Children.Length because Children is an IEnumerable.

File: src/Compilers/Core/Portable/Operations/Operation.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/Operation.cs]: Fixed.

**Pull request merged**
